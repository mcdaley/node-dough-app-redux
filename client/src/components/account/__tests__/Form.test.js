//-----------------------------------------------------------------------------
// client/src/components/account/__tests__/Form.test.js
//-----------------------------------------------------------------------------
import React        from 'react'
import {
  render,
  cleanup,
  fireEvent,
  wait,
}                   from '@testing-library/react'

import AccountForm  from '../Form'

describe('AccountForm', () => {
  afterEach(cleanup)

  it('Takes a snapshot', () => {
    const { asFragment } = render(<AccountForm />)
    
    expect(asFragment(<AccountForm />)).toMatchSnapshot()
   })

  it('Renders the form to create a new account', () => {
    const { getByLabelText, getByPlaceholderText } = render(<AccountForm />)

    expect(getByLabelText('Account Nickname')).toHaveTextContent('')
    expect(getByLabelText('Financial Institute')).toHaveTextContent('')
    expect(getByLabelText('Account Type')).toHaveTextContent('Checking')
    expect(getByPlaceholderText('Enter account balance')).toHaveTextContent('')
  })

  it('Requires values for account nickname and financial institute', async () => {
    const handler       = jest.fn(e => e.preventDefault())
    const { getByTestId, getByText } = render(<AccountForm onSubmit={handler} />)

    fireEvent.click(getByText('Save'))

    await wait(() => fireEvent.click(getByText('Save')) )

    expect(getByTestId('nickname-error')).toHaveTextContent('Nickname is required')
    expect(getByTestId('fi-error')).toHaveTextContent('Financial Institute is required')
  })

  it('Requires account balance to be a number', async () => {
    const handler = jest.fn(e => e.preventDefault())
    const { 
      getByText, 
      getByTestId,
      getByPlaceholderText }  = render(<AccountForm onSubmit={handler} />)

    await wait( () => {
      fireEvent.change(getByPlaceholderText('Enter account balance'), {
        target: {value: 'bad'}
      })
    })
    
    await wait(() => fireEvent.click(getByText('Save')) )

    expect(getByTestId('balance-error')).toHaveTextContent('Balance must be a number')
  })
})