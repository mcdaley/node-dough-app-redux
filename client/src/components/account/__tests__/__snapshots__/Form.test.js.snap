// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm Takes a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class=""
      data-testid="account-form"
      novalidate=""
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="nickname"
        >
          Account Nickname
        </label>
        <input
          class="form-control"
          id="nickname"
          name="nickname"
          placeholder="Enter account nickname"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
          data-testid="nickname-error"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="financialInstitute"
        >
          Financial Institute
        </label>
        <input
          class="form-control"
          id="financialInstitute"
          name="financialInstitute"
          placeholder="Enter financial institute"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
          data-testid="fi-error"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="accountType"
        >
          Account Type
        </label>
        <select
          class="form-control"
          id="accountType"
          name="accountType"
        >
          <option
            value="Checking"
          >
            Checking
          </option>
          <option
            value="Savings"
          >
            Savings
          </option>
          <option
            value="Credit Card"
          >
            Credit Card
          </option>
        </select>
      </div>
      <div
        class="form-group"
      >
        <div
          class="form-row"
        >
          <div
            class="col"
          >
            <label
              class="form-label"
            >
              Account Balance
            </label>
            <input
              class="form-control"
              id="balance"
              name="balance"
              placeholder="Enter account balance"
              required=""
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
              data-testid="balance-error"
            />
          </div>
          <div
            class="col"
          >
            <label
              class="form-label"
            >
              As of Date
            </label>
            <div>
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class="form-control"
                    id="asOfDate"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        class="btn btn-primary"
        style="margin-right: 1.0rem;"
        type="submit"
      >
        Save
      </button>
      <button
        class="btn btn-secondary"
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
